File: Binary_Search_Trees/BinarySearchTreesDS.cpp
#include <iostream>

class Node {
public:
    int data;
    Node* left;
    Node* right;

    Node(int value) {
        data = value;
        left = nullptr;
        right = nullptr;
    }
};

class BinarySearchTree {
private:
    Node* root;

public:
    BinarySearchTree() {
        root = nullptr;
    }

    void insert(int value) {
        root = insertRecursive(root, value);
    }

    Node* insertRecursive(Node* root, int value) {
        if (root == nullptr) {
            return new Node(value);
        }

        if (value < root->data) {
            root->left = insertRecursive(root->left, value);
        } else if (value > root->data) {
            root->right = insertRecursive(root->right, value);
        }

        return root;
    }

    bool search(int value) {
        return searchRecursive(root, value);
    }

    bool searchRecursive(Node* root, int value) {
        if (root == nullptr) {
            return false;
        }

        if (value == root->data) {
            return true;
        } else if (value < root->data) {
            return searchRecursive(root->left, value);
        } else {
            return searchRecursive(root->right, value);
        }
    }

    void inorderTraversal() {
        inorderTraversalRecursive(root);
    }

    void inorderTraversalRecursive(Node* root) {
        if (root != nullptr) {
            inorderTraversalRecursive(root->left);
            std::cout << root->data << " ";
            inorderTraversalRecursive(root->right);
        }
    }


    // function to construct Balanced BST from sorted array

    // Input: 1 2 3 4 5 6 7
    // Output: 4 2 6 1 3 5 7 (Inorder traversal of the constructed BST)
    // Tree:        4
    //            /     \
    //          2         6
    //         /  \     /  \
    //       1   3     5    7

    Node* sortedArrayToBST(int arr[], int start, int end) {
        if (start > end) {
            return nullptr;
        }

        int mid = (start + end) / 2;
        Node* root = new Node(arr[mid]);

        root->left = sortedArrayToBST(arr, start, mid - 1);
        root->right = sortedArrayToBST(arr, mid + 1, end);

        return root;
    }

    Node* sortedArrayToBST(int arr[], int n) {
       
        return sortedArrayToBST(arr, 0, n-1);
    }


};

int main() {
    BinarySearchTree bst;

    bst.insert(50);
    bst.insert(30);
    bst.insert(20);
    bst.insert(40);
    bst.insert(70);
    bst.insert(60);
    bst.insert(80);

    std::cout << "Inorder Traversal: ";
    bst.inorderTraversal();
    std::cout << std::endl;

    // int searchValue = 40;
    // if (bst.search(searchValue)) {
    //     std::cout << searchValue << " found in the tree." << std::endl;
    // } else {
    //     std::cout << searchValue << " not found in the tree." << std::endl;
    // }


    int arr[] = {1, 2, 3, 4, 5, 6, 7};
    int n = sizeof(arr) / sizeof(arr[0]);
    Node* root = bst.sortedArrayToBST(arr,n);
    bst.inorderTraversalRecursive(root);






    return 0;
}

File: Binary_Search_Trees/BinarySearchTreesDS.py
class Node:
    def __init__(self, key):
        self.key = key
        self.left = None
        self.right = None

class BinarySearchTree:
    def __init__(self):
        self.root = None

    def insert(self, key):
        if self.root is None:
            self.root = Node(key)
        else:
            self._insert_recursive(self.root, key)

    def _insert_recursive(self, node, key):
        if key < node.key:
            if node.left is None:
                node.left = Node(key)
            else:
                self._insert_recursive(node.left, key)
        else:
            if node.right is None:
                node.right = Node(key)
            else:
                self._insert_recursive(node.right, key)

    def search(self, key):
        return self._search_recursive(self.root, key)

    def _search_recursive(self, node, key):
        if node is None or node.key == key:
            return node
        if key < node.key:
            return self._search_recursive(node.left, key)
        return self._search_recursive(node.right, key)

    def delete(self, key):
        self.root = self._delete_recursive(self.root, key)

    def _delete_recursive(self, node, key):
        if node is None:
            return node
        if key < node.key:
            node.left = self._delete_recursive(node.left, key)
        elif key > node.key:
            node.right = self._delete_recursive(node.right, key)
        else:
            if node.left is None:
                return node.right
            elif node.right is None:
                return node.left
            min_node = self._find_min(node.right)
            node.key = min_node.key
            node.right = self._delete_recursive(node.right, min_node.key)
        return node

    def _find_min(self, node):
        current = node
        while current.left is not None:
            current = current.left
        return current

    def inorder_traversal(self):
        self._inorder_recursive(self.root)

    def _inorder_recursive(self, node):
        if node is not None:
            self._inorder_recursive(node.left)
            print(node.key, end=" ")
            self._inorder_recursive(node.right)
   
   
            
            
    #  Function to check given tree is BST or not

    #  Tree:        50
    #             /     \
    #           30       70
    #          /  \     /  \
    #       20   40   60   80

    #  Output: Given tree is a BST
    
    def isBST(self, root, l = None, r = None):
        if root is None:
            return True
        if l is not None and root.key <= l.key:
            return False
        if r is not None and root.key >= r.key:
            return False
        return self.isBST(root.left, l, root) and self.isBST(root.right, root, r)
            
            
           

bst = BinarySearchTree()

# Insert nodes
bst.insert(50)
bst.insert(30)
bst.insert(20)
bst.insert(40)
bst.insert(70)
bst.insert(60)
bst.insert(80)

# Search for a node
search_key = 40
result = bst.search(search_key)
if result:
    print(f"Node with key {search_key} found!")
else:
    print(f"Node with key {search_key} not found!")

            # Delete a node
# delete_key = 30
# bst.delete(delete_key)
# print(f"Node with key {delete_key} deleted!")

# Inorder traversal
print("Inorder Traversal:")
bst.inorder_traversal()


print("\nGiven tree is a BST") if bst.isBST(bst.root) else print("Given tree is not a BST")